torch/csrc/autograd/saved_variable.cpp

```diff
@@ -90,6 +98,9 @@ void SavedVariable::save_common_metadata(const Variable& data) {
   }
 }
 
+std::unique_ptr<SavedVariableHooks> SavedVariable::get_default_hooks() {
+  return Engine::get_default_engine().get_default_saved_variable_hooks();
+}
```

Comment:
> I'm using unique_ptr because it makes it easier to guarantee that the underlying PyObject* pack_hook and unpack_hook are correctly increfed and decrefed. This does create a new SavedVariableHooks object with its own pair of hooks for each SavedVariable, but we need that in order for the hooks to survive when default hooks are unset but the SavedVariable is still alive.

Shouldn't the way to model this be std::shared_ptr? In C++, on construction of the `std::shared_ptr<SavedVarableHooks>` we grab the gil and incref the pack_hook and unpack_hook. When the shared_ptr gets deleted, we grab the gil and decref the pack_hook and unpack_hook.

`std::unique_ptr<SavedVariableHooks>` means that every time we construct it, it allocates a SavedVariableHooks on the heap which is something we really want to avoid.

Reply:
That being said, this is orthogonal to this refactor so it doesn't have to come here in the stack, but I think we should really file an issue and/or fix it

Reply:
I see your point. I will open an issue about it

Reply:
@zou3519: After looking into it further with @albanD, it turns out we do need each `SavedVariable` to have its own `PySavedVariableHooks` because that struct stores the `SavedVariable`-specific `PyObject* data_`.
We could overcome this by storing `PyObject* data_` directly into `SavedVariable` but that limits the extensibility of this feature for when we want to implement hooks in C++